daemon/api_blob.go

```diff
@@ -0,0 +1,67 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"net/http"
+
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+var snapBlobCmd = &Command{
+	Path:     "/v2/snaps/{name}/blob",
+	UserOK:   true,
+	PolkitOK: "io.snapcraft.snapd.manage",
+	GET:      getSnapBlob,
+}
+
+func getSnapBlob(c *Command, r *http.Request, user *auth.UserState) Response {
+	vars := muxVars(r)
+	name := vars["name"]
+
+	st := c.d.overlord.State()
+	st.Lock()
+	var snapst snapstate.SnapState
+	var info *snap.Info
+	err := snapstate.Get(st, name, &snapst)
+	if err == nil {
+		info, err = snapst.CurrentInfo()
+	}
+	st.Unlock()
+	switch err {
+	case nil:
+		// ok
+	case state.ErrNoState:
+		return SnapNotFound(name, err)
+	default:
+		return InternalError("cannot download blob for snap %q: %v", name, err)
+	}
+	if !snapst.Active {
```

Comment:
Why cannot we get a blob of an inactive snap? Is there a technical reason or a design reason behind this?

Reply:
We don't want to expose more than the user could get by copying things that are already mounted

Reply:
But inactive snaps _are_ mounted. They are not exposed as far as systemd units or launchers are concerned but they are definitely not invisible in SNAP_MOUNT_DIR.

Reply:
Yes, that bit is more about not exposing things that aren't needed that we're not sure will remain the way they are in the future.
That is: inactive snaps are mounted, but tomorrow we might stop doing that (we have people asking for that, with good reasons for it). And nothing inherently requires that we have them mounted. And the intended user of the api doesn't need them. So, safest to leave them out.

Reply:
I think we should treat "snap download" as a cache. If you have it on disk, just get that. Putting arbitrary restrictions won't stop anyone from "sudo cp" things out of /var/.